File: App/src/Adaptive.ReactiveTrader.Client/UI/SpotTiles/OneWayPriceViewModel.cs
using System;
using System.Reactive.Linq;
using System.Windows;
using System.Windows.Input;
using Adaptive.ReactiveTrader.Client.Domain.Models;
using Adaptive.ReactiveTrader.Shared.UI;
using log4net;
using PropertyChanged;

namespace Adaptive.ReactiveTrader.Client.UI.SpotTiles
{
    [ImplementPropertyChanged]
    public class OneWayPriceViewModel : ViewModelBase, IOneWayPriceViewModel
    {
        private readonly ISpotTilePricingViewModel _parent;
        private static readonly ILog Log = LogManager.GetLogger(typeof(OneWayPriceViewModel));

        private readonly DelegateCommand _executeCommand;
        private IExecutablePrice _executablePrice;

        public Direction Direction { get; private set; }
        public string BigFigures { get; private set; }
        public string Pips { get; private set; }
        public string TenthOfPip { get; private set; }
        public bool IsExecuting { get; private set; }
        public bool IsStale { get; private set; }
        
        public OneWayPriceViewModel(Direction direction, ISpotTilePricingViewModel parent)
        {
            _parent = parent;
            Direction = direction;

            _executeCommand = new DelegateCommand(OnExecute, CanExecute);
        }

        #region ExecuteCommand
        public ICommand ExecuteCommand { get { return _executeCommand; } }

        private bool CanExecute()
        {
            return _executablePrice != null && !IsExecuting;
        }

        private void OnExecute()
        {
            long notional;
            if (!long.TryParse(_parent.Notional, out notional))
            {
                // TODO handle notional validation properly
                return;
            }
            IsExecuting = true;
            _executablePrice.Execute(notional, _parent.DealtCurrency)
                .ObserveOnDispatcher()
                .Subscribe(OnExecuted,
                    OnExecutionError);    
        }

        private void OnExecutionError(Exception exception)
        {
            if (exception is TimeoutException)
            {
                MessageBox.Show("No response was received from the server, the execution status is unknown.\nPlease contact your sales representative.", "Execution timeout", MessageBoxButton.OK, MessageBoxImage.Error);
                Log.Error("Trade execution request timed out.");
            }
            else
            {
                MessageBox.Show("An error occured while processing the trade request.", "Execution error", MessageBoxButton.OK, MessageBoxImage.Error);
                Log.Error("An error occured while processing the trade request.", exception);
            }
            IsExecuting = false;
        }

        #endregion

        public void OnPrice(IExecutablePrice executablePrice)
        {
            _executablePrice = executablePrice;

            var formattedPrice = PriceFormatter.GetFormattedPrice(_executablePrice.Rate,
                executablePrice.Parent.CurrencyPair.RatePrecision, executablePrice.Parent.CurrencyPair.PipsPosition);

            BigFigures = formattedPrice.BigFigures;
            Pips = formattedPrice.Pips;
            TenthOfPip = formattedPrice.TenthOfPip;
            IsStale = false;

            _executeCommand.RaiseCanExecuteChanged();
        }

        public void OnStalePrice()
        {
            _executablePrice = null;

            BigFigures = string.Empty;
            Pips = string.Empty;
            TenthOfPip = string.Empty;
            IsStale = true;

            _executeCommand.RaiseCanExecuteChanged();
        }

        private void OnExecuted(ITrade trade)
        {
            Log.Info("Trade executed");
            _parent.OnTrade(trade);
            IsExecuting = false;
        }
    }

    public enum PriceMovement
    {
        None,
        Down,
        Up
    }
}

File: src/Adaptive.ReactiveTrader.Client.JS/UI/SpotTiles/OneWayPriceViewModel.ts
class OneWayPriceViewModel implements IOneWayPriceViewModel {
    private _parent: ISpotTileViewModel;
    private _executablePrice: IExecutablePrice;

    constructor(parent: ISpotTileViewModel, direction: Direction) {
        this._parent = parent;
        this.direction = direction == Direction.Buy ? "BUY" : "SELL";

        this.bigFigures = ko.observable("");
        this.pips = ko.observable("");
        this.tenthOfPips = ko.observable("");
        this.isExecuting = ko.observable(false);
        this.isStale = ko.observable(true);
    }

    onExecute(): void {
        this.isExecuting(true);

        this._executablePrice.execute(this._parent.notional(), this._parent.dealtCurrency)
            .subscribe(
                trade => this.onExecuted(trade),
                ex => this.onExecutionError(ex));
    }

    onPrice(executablePrice: IExecutablePrice): void {
        this._executablePrice = executablePrice;

        var formattedPrice = PriceFormatter.getFormattedPrice(executablePrice.rate,
            executablePrice.parent.currencyPair.ratePrecision, executablePrice.parent.currencyPair.pipsPosition);

        this.bigFigures(formattedPrice.bigFigures);
        this.pips(formattedPrice.pips);
        this.tenthOfPips(formattedPrice.tenthOfPips);
        this.isStale(false);
    }

    onStalePrice(): void {
        this._executablePrice = null;
        this.bigFigures("");
        this.pips("");
        this.tenthOfPips("");
        this.isStale(true);
    }

    private onExecuted(trade: ITrade) {
        console.log("Trade executed.");
        this._parent.onTrade(trade);
        this.isExecuting(true);
    }

    private onExecutionError(ex: Error) {
        // TODO
        console.error(ex);
    }

    direction: string;
    bigFigures: KnockoutObservable<string>;
    pips: KnockoutObservable<string>;
    tenthOfPips: KnockoutObservable<string>;
    isExecuting: KnockoutObservable<boolean>;
    isStale: KnockoutObservable<boolean>;
} 